standard/classes.md

```diff
@@ -434,12 +433,16 @@ Each *type_parameter_constraints_clause* consists of the token `where`, followed
 
 The list of constraints given in a `where` clause can include any of the following components, in this order: a single primary constraint, one or more secondary constraints, and the constructor constraint, `new()`.
 
-A primary constraint can be a class type, the ***reference type constraint*** `class`, the ***value type constraint*** `struct`, or the ***unmanaged type constraint*** `unmanaged`.
+A primary constraint can be a class type, the ***reference type constraint*** `class`, the ***value type constraint*** `struct`, the ***not null constraint*** `notnull` or the ***unmanaged type constraint*** `unmanaged`.
 
-A secondary constraint can be a *type_parameter* or *interface_type*.
+A secondary constraint can be an *interface_type* or *type_parameter*, optionally followed by a *nullable_type_attribute*.
 
 The reference type constraint specifies that a type argument used for the type parameter shall be a reference type. All class types, interface types, delegate types, array types, and type parameters known to be a reference type (as defined below) satisfy this constraint.
 
+The class type, reference type constraint, and secondary constraints can include the nullable type attribute. The nullability of the type argument need not match the nullability of the type parameter. The compiler may issue a warning if the nullability of the type parameter doesn't match the nullability of the type argument.
```

Comment:
I suspect "type parameter" should be "constraint" here.

Reply:
We're unsure whether `class?` is intended to mean "definitely nullable" or "might be nullable". Current description suggests that using `string` for a type parameter of `T : class?` could issue a warning.

There are three things users might want to specify:

- "T should be a non-nullable reference type"
- "T should be a nullable reference type"
- "T should be any reference type"

Currently we only have two constraints for those three cases, which is far from ideal.

Reply:
The 2nd isn't actually a constraint. To say "T should be a nullable reference type" you say "T is a non-nullable reference type" and then use `T?` consistently at all uses.

Reply:
Clarification from Mads: `class?` is intended to mean "might be nullable". If you want "definitely nullable" you don't specify it as a constraint, you constrain T to be non-nullable, but use T? where you want to allow a null value.

Reply:
Also, work out what `T?` means if `T` is the type parameter, and `string?` (or some other nullable reference type) is the type argument.

Reply:
Also note that `typeof(T?)` is an error if `T` is known to be a reference type?

Reply:
> Also note that `typeof(T?)` is an error if `T` is known to be a reference type?

Yes, and it's an error in 8 if `T` is unconstrained. It's valid *but useless* in 9+ when `T` is unconstrained (`typeof(T)` and `typeof(T?)` always return the same Type). We might want to file a bug about this.

Reply:
> > Also note that `typeof(T?)` is an error if `T` is known to be a reference type?
> 
> Yes, and it's an error in 8 if `T` is unconstrained. It's valid _but useless_ in 9+ when `T` is unconstrained (`typeof(T)` and `typeof(T?)` always return the same Type). We might want to file a bug about this.

These changes are added in expressions.md, in section 12.8.12 (The `typeof` operator)